sbi/diagnostics/tarp.py

```diff
@@ -0,0 +1,407 @@
+                 sampling a trained posterior `num_samples` times. Multiple
+                 (posterior) samples for one observation are encouraged.
+        theta: The true parameter value theta. Theta is expected to
+                 have shape ``(num_sims, num_dims)``.
+        references: the reference points to use for the coverage regions, with
+                shape ``(1, num_sims, num_dims)``, or ``None``.
+                If ``None``, then reference points are chosen randomly from
+                the unit hypercube over the parameter space given by theta.
+                In other words, reference samples are drawn from the
+                following ``Uniform(low=theta.min(dim=-1),high=theta.max(dim=-1))``.
+        distance: the distance metric to use when computing the distance.
+                Should be a callable function that accepts two tensors and
+                computes the distance between them, e.g. given two tensors
+                of shape ``(batch, 3)`` and ``(batch,3)``, this function should
+                return ``(batch,1)`` distance values.
+                Possible values: ``sbi.utils.metrics.l1`` or
+                ``sbi.utils.metrics.l2``. ``l2`` is the default.
+        num_bins: number of bins to use for the credibility values.
+                If ``None``, then ``num_sims // 10`` bins are used.
+        do_norm : whether to normalize parameters before coverage test
+                (Default = True)
+        rng_seed : whether to set the seed of torch.random, no seed is set
+                if None is received
+                (Default = None)
+
+    Returns:
+        ecp: Expected coverage probability (``ecp``), see equation 4 of the paper
+        alpha: credibility values, see equation 2 of the paper
+
+    """
+    # TARP assumes that the predicted thetas are sampled from the "true"
+    # PDF num_samples times
+    theta = theta.detach() if len(theta.shape) != 2 else theta.detach().unsqueeze(0)
+    samples = samples.detach()
+
+    assert (
+        theta.shape[-2:] == samples.shape[-2:]
+    ), f"shapes of theta {theta.shape[-2:]} and samples {samples.shape[-2:]} do not fit"
+
+    num_samples = samples.shape[0]  # samples per simulation
+
+    if num_bins is None:
+        num_bins = samples.shape[-2] // 10
+
+    if do_norm:
+        lo = torch.min(
+            theta, dim=-2, keepdims=True
+        ).values  # min along samples.shape[-2]
+        hi = torch.max(
+            theta, dim=-2, keepdims=True
+        ).values  # max along samples.shape[-2]
+        samples = (samples - lo) / (hi - lo + 1e-10)
+        theta = (theta - lo) / (hi - lo + 1e-10)
+
+    references = _check_references(theta, references)
+    assert len(references.shape) == len(
+        samples.shape
+    ), f"shape mismatch of references {references.shape} and samples {samples.shape}"
+
+    # distances between references and samples
+    sample_dists = distance(references.expand(num_samples, -1, -1), samples)
```

Comment:
the expand is not needed, it will return the distance for each entry in the leading dimension of `samples`, as long as the last two dims are matching. 
```suggestion
    sample_dists = distance(references, samples)
```